docs/OpenSSLSupport.md

```diff
@@ -0,0 +1,196 @@
+pem.h | Yes | - |
+pem2.h | Yes | - |
+pemerr.h | Yes | - |
+pkcs12.h | Yes | - |
+pkcs12err.h | Yes | - |
+pkcs7.h | Yes | - |
+pkcs7err.h | Yes | - |
+rand.h | Partial | EGD is disabled by default. The default method (RAND_OpenSSL) does not work because the depending `rdtsc` instruction is not supported by SGXv1. Refer to [How to use RAND APIs](#how-to-use-rand-apis) for more detail. |
+rand_drbg.h | Partial | OE by default does not depend on the default rand method. Therefore, rand_drbg APIs are supported but have no impact on rand APIs. The drbg test is disabled. Refer to the [unsuppored test list](/tests/openssl/tests.unsupported) for more detail. |
+randerr.h | Yes | - |
+rc2.h | Yes | - |
+rc4.h | Yes | - |
+rc5.h | No | RC5 is disabled by default (header is present). |
+ripemd.h | Yes | - |
+rsa.h | Yes | - |
+rsaerr.h | Yes | - |
+safestack.h | Yes | - |
+seed.h | Yes | - |
+sha.h | Yes | - |
+srp.h | Yes | - |
+srtp.h | Yes | - |
+ssl.h | Partial | SSL2 and SSL3 methods are disabled. Heartbeats extension is disabled by defailt. |
+ssl2.h | Yes | - |
+ssl3.h | Yes | - |
+sslerr.h | Yes | - |
+stack.h | Yes | - |
+store.h | Yes | - |
+storeerr.h | Yes | - |
+symhacks.h | Yes | - |
+tls1.h | Partial | Heartbeats extension is disabled by default. |
+ts.h | Yes | - |
+tserr.h | Yes | - |
+txt_db.h | Yes | - |
+ui.h | No | Built with OPENSSL_NO_UI_CONSOLE. |
+uierr.h | Yes | - |
+whrlpool.h | Yes | - |
+x509.h | Yes | - |
+x509_vfy.h | Yes | - |
+x509err.h | Yes | - |
+x509v3.h | Yes | - |
+x509v3err.h | Yes | - |
+
+# How to use RAND APIs
+
+Given that the default OpenSSL RAND method relies on the `rdtsc` instruction that is not supported by SGXv1 enclaves,
+OE currently does not support RAND APIs if users try to use them directly (by default, the RAND APIs depend on the default
+OpeSSL RAND method). To enable RAND APIs, OE recommends to use the OpenSSL RDRAND engine, which explicitly replaces the
+default RAND method with the `rdrand`-based method. That is, the RAND APIs will obtain the random bytes directly using the
+`rdrand` instruction. To opt-in the RDRAND engine, see the following example.
+
+```
+void enc_rand()
+{
+    int data = 0;
+    ENGINE* eng = NULL;
+
+    /* Initialize and opt-in the RDRAND engine. */
+    ENGINE_load_rdrand();
+    eng = ENGINE_by_id("rdrand");
+    if (eng == NULL)
+    {
+        goto done;
+    }
+
+    if (!ENGINE_init(eng))
+    {
+        goto done;
+    }
+
+    if (!ENGINE_set_default(eng, ENGINE_METHOD_RAND))
+    {
+        goto done;
+    }
+
+    /* Now RAND APIs are available. */
+
+    /* Test the RAND_bytes API. */
+    if (!RAND_bytes((unsigned char*)&data, sizeof(data)))
+    {
+        goto done;
+    }
+
+    printf("RAND_bytes() %d\n", data);
+
+done:
+
+    /* cleanup to avoid memory leak. */
+    ENGINE_finish(eng);
+    ENGINE_free(eng);
+    ENGINE_cleanup();
+
+    return;
+}
+```
+
+Note that the code snippet for the RDRAND engine opt-in is required to use not only RAND APIs
```

Comment:
Follow-up: I would strongly suggest restructuring the document so that this is not buried at the bottom. 

* Understanding that much of the OpenSSL library is broken unless the developer writes some code to use the rdrand engine is something that needs to be super clear to a developer before they even worry about specific bits of functionality.
* Talking about the rand handling provides context for explaining other choices like the `--with-rand-seed=none` option.

There's enough here that I would split this into 2 docs that reference each other
* a how-to doc with instructions on libraries to link to use openssl and how to initialize (including rand engine setting) 
* a reference doc for supported functionality (this doc) with a link back to the how-to doc. 

For this to move out of experimental, I think it's not acceptable for the user to have to go and write code just to get what is usually expected as a baseline bit of functionality. As the liboecryptoopenssl piece gets implemented, we are going to have to implement an initialization for OE use anyway, and making that available as part of the aggregated libopenssl would be useful. (e.g. `oe_openssl_init_ssl`/`oe_open_ssl_init_crypto` even though that's a little bit of a throwback to < 1.1.0 API, it's still better than making them write a whole function for basic use of openssl.)

Reply:
I restructured the doc by putting the rand API section before the list API support for now (and mention the current support is experimental). Given the goal is not requiring users to write those initialization code, the current content would just be temporary for the experimental support. I'd like to split the doc into two when we're ready to move out of experimental.